core/trino-parser/src/main/antlr4/io/trino/jsonpath/JsonPath.g4

```diff
@@ -0,0 +1,200 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+grammar JsonPath;
+
+tokens {
+    DELIMITER
+}
+
+path
+    : pathMode pathExpression EOF
+    ;
+
+pathMode
+    : LAX
+    | STRICT
+    ;
+
+pathExpression
+    : accessorExpression                                                    #expressionDefault
+    | sign=('+' | '-') pathExpression                                       #signedUnary
+    | left=pathExpression operator=('*' | '/' | '%') right=pathExpression   #binary
+    | left=pathExpression operator=('+' | '-') right=pathExpression         #binary
+    ;
+
+accessorExpression
+    : pathPrimary                                               #accessorExpressionDefault
+    | accessorExpression '.' identifier                         #memberAccessor
+    | accessorExpression '.' stringLiteral                      #memberAccessor
+    | accessorExpression '.' '*'                                #wildcardMemberAccessor
+    | accessorExpression '[' subscript (',' subscript)* ']'     #arrayAccessor
+    | accessorExpression '[' '*' ']'                            #wildcardArrayAccessor
+    | accessorExpression '?' '(' predicate ')'                  #filter
+    | accessorExpression '.' TYPE '(' ')'                       #typeMethod
+    | accessorExpression '.' SIZE '(' ')'                       #sizeMethod
+    | accessorExpression '.' DOUBLE '(' ')'                     #doubleMethod
+    | accessorExpression '.' CEILING '(' ')'                    #ceilingMethod
+    | accessorExpression '.' FLOOR '(' ')'                      #floorMethod
+    | accessorExpression '.' ABS '(' ')'                        #absMethod
+    | accessorExpression '.' DATETIME '(' stringLiteral?  ')'   #datetimeMethod
+    | accessorExpression '.' KEYVALUE '(' ')'                   #keyValueMethod
+    ;
+
+identifier
+    : IDENTIFIER
+    | nonReserved
+    ;
+
+subscript
+    : singleton=pathExpression
+    | from=pathExpression TO to=pathExpression
+    ;
+
+pathPrimary
```

Comment:
leafExpression?

Reply:
`...Primary` is a naming convention we use in the `SqlBase` grammar to denote this abstraction.
In fact, it is not a leaf, but rather kind of a syntactically distinguished unit.